docs/question.md

```diff
@@ -438,12 +445,44 @@ Example of valid HTML:
+
+### Custom Question Best Practices
+
+Although custom questions usually don't use the grading functions from individual elements, it is _highly_ recommended that built-in elements are used for student input. Parsed student answers are present in the `data["submitted_answers"]` dictionary.
+
+Any custom grading function for the whole question should set `data["score"]` as a value between 0.0 and 1.0, which will be the final score for the given question. If a custom grading function is only grading a specific part of a question, the grading function should set the corresponding dictionary entry in `data["partial_scores"]` and then recompute the final `data["score"]` value for the whole question. Alternatively, if you would prefer not to show score badges for individual parts, you may unset the dictionary entries in `data["partial_scores"]`.
+
+To set custom feedback, the grading function should set the corresponding entry in the `data["feedback"]` dictionary. These feedback entries are passed in when rendering the `question.html`, which can be accessed by using the mustache prefix `{{feedback.}}`. See the [above custom question](#Question-server.py) or [this demo question](https://github.com/PrairieLearn/PrairieLearn/tree/master/exampleCourse/questions/demo/custom/gradeFunction) for examples of this. Note that the feeback set in the `data["feedback"]` dictionary is meant for use by custom grader code in a `server.py` file, while the feedback set in `data["partial_scores"]` is meant for use by element grader code.
+
+For generated floating point answers, it's important to round values before displaying them to the student, as this could cause unexpected behavior when students perform calculations. For example, the following is problematic:
+
+```python
+def generate(data):
+    a = 33.33337
+    b = 33.33333
+    data["params"]["a_for_student"] = f'{a:.2f}'
+    data["params"]["b_for_student"] = f'{a:.2f}'
+    data["correct_answers"]["c"] = a - b
+```
+
+The rounding should be done before the format strings are set, as in the following example:
+
+```python
+def generate(data):
+  a = np.round(33.33337, 2)
+  b = np.round(33.33333, 2)
+  data["params"]["a_for_student"] = f'{a:.2f}'
+  data["params"]["b_for_student"] = f'{b:.2f}'
+  data["correct_answers"]["c"] = a - b
+```
+
+Similarly, for grading functions involving floating point numbers, _avoid exact comparisons with `==`._ Floating point calculations in Python introduce error, and comparisons with `==` might unexpectedly fail. Instead, the function [`math.isclose`](https://docs.python.org/3/library/math.html#math.isclose) can be used, as it performs comparisons within given tolerance values. The functions `pl.is_correct_scalar_ra`, `pl.is_correct_scalar_sf`, and `pl.is_correct_scalar_dd` can be used to perform more specialized comparisons.
```

Comment:
Consider linking to the `prairielearn` module or providing inline examples here, or maybe even both. If I didn't know about `import pl from prairielearn`, I wouldn't know where to find these functions or how to use them.

Reply:
I think that makes sense. There are some example usages I can link to with these functions, and each of them is documented so that should be enough for people to get started. I'll probably also drop the `pl.*` import prefix and just state where they can be found.

Reply:
Cleaned this description up some and included links to the functions in `prairielearn.py`. Also, added docstrings to those functions describing what they do.

Reply:
Yeah, let's just link to the file. The specific line numbers are likely to become out-of-date almost immediately.